### PR TITLE
Move most instance deriving into the type checker

### LINKED ISSUE #3453: Derive newtype hangs on self-recursive type synonym
Minimal repro:
```
module Test where

import Data.Newtype (class Newtype)

type S = S
newtype Z = Z S
derive instance newtypeZ :: Newtype Z _
```

Without the `derive instance` line it fails quickly, as expected, but with that line, the compiler hangs for a long time. It seems like it does not check if the type it is generating actually makes sense.

### LINKED ISSUE #3510: Incorrect error when trying to derive instances for type synonyms
For example:
```purescript
module Main where
import Prelude (class Eq)
type T = {}
derive instance eqT :: Eq T
```
Currently gives you
> `CannotFindDerivingType`: "Cannot derive a type class instance, because the type declaration for T could not be found."

Perhaps 

> `TypeSynonymInstance`: "Type class instances for type synonyms are disallowed."

would be more appropriate?

### LINKED ISSUE #4086: Make typeclass and deriving desugaring part of type checking
## Summary

Make certain kind information available sooner in the compilation pipeline, such that other desirable features become unblocked or easier to do.

## Motivation

See [kl0tl's comment and natefaubion's following comment in #3426](https://github.com/purescript/purescript/issues/3426#issuecomment-651082718). Fixing this would unblock a few things:
- #3426
- #4084 
- potentially other things I'm not aware of

Since this is far outside my expertise, I'm not able to provide a concrete proposal or example. However, this at least provides a place to discuss this further.

### LINKED ISSUE #4105: Type synonym error on Identity: Type synonym ... is partially applied.
## Description

I wanted to update my project from PureScript 0.13 to 0.14. Getting the error message:

```txt
Type synonym Misc.Template is partially applied.
Type synonyms must be applied to all of their type arguments.
```

## To Reproduce

Misc.purs:

```purescript
module Misc where
  
type Template col = { bio :: col String }

type Identity a = a

type Patch = Template Identity
```

+ Other.purs:
```purescript
module Other where

import Prelude

import Misc 

newtype UpdateDto = UpdateDto Patch

derive instance eqUpdateDto :: Eq UpdateDto
```
- It can’t be triggered in one file.
- It has one of those error positions where it’s caused by the existence of Other.purs, but identified in Misc.purs
- It can be worked around by inlining the Identity synonym usage (newtype UpdateDto = UpdateDto (Template Identity))
- Possibly related to  #4101.

## Expected behavior

No error like in version 0.13.

## Additional context

[discourse](https://discourse.purescript.org/t/update-purescript-0-13-0-14-type-synonyms/2373)
## PureScript version

0.14.2

### LINKED ISSUE #4200: KindsDoNotUnify error when deriving Newtype from alias
## Description

The case, seems like a bug: 
- there is a parameterized data type and an alias for this data type (in practice defined in different modules)
- in another module deriving `Newtype` instance for a `newtype` that contains mentioned alias type
- the compiler gives an error `KindsDoNotUnify`

## To Reproduce

The minimal example to reproduce requires two modules:

```purs
module Test.A where

import Test.B (TAlias)
import Data.Newtype (class Newtype)

newtype NewA a = NewA (TAlias Int)

derive instance Newtype (NewA a) _
```

```purs
module Test.B where

data T :: forall m. m -> Type
data T msg = E

type TAlias :: forall k. k -> Type
type TAlias msg =
  T msg -- here is an error
```

The error:
```
  Could not match kind

    Type

  with kind

    k


while checking that type Int
  has kind k
while inferring the kind of T @k Int
while inferring the kind of Newtype (NewA a) (T @k Int)
in type class instance

  Data.Newtype.Newtype (NewA a)
                       (T @k Int)
PureScript(KindsDoNotUnify)
```

## Expected behavior

Should be no error.

## Additional context

It starts working if to replace `TAlias's` kind signature with `Type -> Type`

## PureScript version

0.14.5

### PR DESCRIPTION
**Pull request by rhendric**
**Description of the change**

Closes #4086. And, as a consequence, fixes #3453, closes #3510, fixes #4105, and fixes #4200.

---

**Checklist:**

- [x] Added a file to CHANGELOG.d for this PR (see CHANGELOG.d/README.md)
- <s>Added myself to CONTRIBUTORS.md (if this is my first contribution)</s>
- [x] Linked any existing issues or proposals that this pull request should close
- <s>Updated or added relevant documentation</s>
- [x] Added a test for the contribution (if applicable)

### CODE REVIEW

**Review by JordanMartinez — APPROVED** [review]
Whew! Got through this. Besides the nitpick indentation, LGTM.

**Comment by JordanMartinez** [comment]
Oh, and there's a merge conflict because your other PR got merged.

File: src/Language/PureScript/TypeChecker/Deriving.hs [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+      unaryClass f = case tys of
+       [ty] -> case unwrapTypeConstructor ty of
+         Just (Qualified (Just mn') tyCon, _, _) | mn == mn' -> do
+           let superclassesDicts = flip map typeClassSuperclasses $ \(Constraint _ superclass _ suTyArgs _) ->
+                 let tyArgs = map (replaceAllTypeVars (zip (map fst typeClassArguments) tys)) suTyArgs
+                 in lam UnusedIdent (DeferredDictionary superclass tyArgs)
+           let superclasses = map mkString (superClassDictionaryNames typeClassSuperclasses) `zip` superclassesDicts
+           App (Constructor nullSourceSpan ctorName) . mkLit . ObjectLiteral . (++ superclasses) <$> f mn tyCon
+         _ -> throwError . errorMessage $ ExpectedTypeConstructor className tys ty
+       _ -> throwError . errorMessage $ InvalidDerivedInstance className tys 1
```
[the remark this answers is not in the document]
Comment (rhendric) [reply src/Language/PureScript/TypeChecker/Deriving.hs]: Fixed!

[the remark this answers is not in the document]
Comment (JordanMartinez) [reply src/Language/PureScript/TypeChecker/Deriving.hs]: Thanks! Feel free to merge this when you think it appropriate!

**Review by JordanMartinez — APPROVED** [review]

**Pull request merged**